include/uart.h

```diff
@@ -125,6 +180,10 @@ struct uart_driver_api {
 	/** Console I/O function */
 	int (*err_check)(struct device *dev);
 
+	/** UART configuration functions */
+	int (*set_cfg)(struct device *dev, struct uart_config *cfg);
+	int (*get_cfg)(struct device *dev, struct uart_config *cfg);
```

Comment:
what is the use case for that one?

Reply:
> Wondering if we should have:
set_baudrate()

I believe this again was discussed, and @tbursztyka suggest to follow how it's done for other comm. interfaces (?, didn't check that myself), and have common structure for all params.

We also should consider a usecase of userspace syscalls (because UART config API of course should be exported to userspace in one or another form). From that PoV, we always should save on # of syscalls. (Can define convenience inline wrappers on top of that).

Reply:
I don't think we need atomic cfg modifier. Just apply a new config via set_cfg as it is done in other buses. It's not like uart cfg has a lot of fields.

I would rename set_cfg to configure btw.

Reply:
> atomic cfg modifier

I'm not sure I understand context for "atomic" here. The question is still whether we need to provide "get cfg" operation *to users*, and that seems to be useful. After all, how one would implement POSIX `stty` for example?

Reply:
I was replying to @galak's idea of set_baudrate() set_parity() etc... I think that is useless.

About the POSIX stty, that means implementing a call in tty place so tty itself could hold the uart config it's using. However, that would mean occupying some more space, better having a get_cfg then. That's a valid use case.

Reply:
Should set_cfg and get_cfg stay as two separate functions, or change it to one function with additional parameter which tells if its write or read operation? i.e.
```
int (*configure)(struct device *dev, struct uart_config *cfg, bool set);
```

Reply:
if purpose of `get` functionality is only to not overwrite existing configuration and we see no need for user to get current uart configuration then maybe each field should have 0 value as 'no change'. you can then easily memset structure to 0's and set field you actually want to change.

With that approach we can stay with `uart_configurate` only and later on if needed add `get` functionality if we really need that.

Reply:
I think this question should discussed with the above: the **flags** in **struct uart_config**.

In my opinion, **get_cfg** API is only used when flags are bitmasks. We should get the current **flags** before change to a new mode such as change the parity, modify the bits relevantd parity, and then write the new **flags** by using **set_cfg** API.

But I prefer using enum for **flags**, in that way, we don't need **get_cfg** API, change the parity at anytime by using APIs such as uart_set_parity...

Reply:
@qianfan-Zhao, as discussed above, there wouldn't be uart_set_parity etc. on subsystem/system implementation level. Thus, we need get_cfg if we want to allow users to define convenience wrappers like that.

Reply:
@pfalcon And what is yours idea about 'u32_t flags' or enum?

Reply:
looks better indeed.
(on spi the reason why there is no configure() is that config is always given to each calls, driver reconfigures if only it's a different one. I2c could have been doing the same actually.)

Reply:
changed to configure() and get_config()